### DISCUSSION TITLE
hyper://dns (mapping DNS to Hyperdrives)

### DISCUSSION THREAD

**Discussion by da2x** [beakerbrowser/beaker#1576] [General]
Any plans to add `hyper://<dns>` akin to `dat://<dns>`? I noticed that the Beaker homepage, Hypercore Protocol’s homepage, and documentation isn’t available as Hyperdrives.

`hyper://e36050c2651b2bd0bd265fc311c122122a5490bc11a17606201964246b69e4ef/` is nice and all but as a human, I prefer typing `blog.hypercore-protocol.org`.

**Comment by bierlingm** [comment]
Could this be done in a peer based way?

Namespaces are usually a point of centralization... And if you don't use a centralized authority managing it, you tend to get all kinds of shenanigans like hogging.

**Comment by pfrazee** [comment]
Yeah we're working on a domain-name solution that we like better than the old method. We have two basic ideas that we're debating...

 1. Directories
 2. Domain-listings

Both of these would be Hyperdrives that maintain `name -> key` listings. The difference is that a "directory" would effectively be search-based and the user could integrate as many directories as they want, while the domain-listings would be a new TLD (or multiple new TLDs) for assigning domain names.

The directories would probably work like `@pfrazee` which, when you enter it, would find me in your configured directories and show my listing in the URL bar autocompletion options, but then redirect to my `hyper://` url.

This is a tough call. Directories centralize much less control since they enable people to configure as many as they like, but they don't work like domains. Domain-listings can work the way domains have previously worked, where I'd be `hyper://pfrazee.p2p` (or whatever the TLD is).

**Reply by pfrazee** [reply]
Update on this--

Maf is currently working on something called HyperDomains or HDNS (Hyper Domain Name System). He's designing it as a Hypertrie structure (not a Hyperdrive). He's investigating whether it'll be possible to store the entries in an encrypted form so that you can't enumerate them, but that's in development.

Using a Hypertrie will give it some interesting properties. First, it's peer-to-peer like everything else in the Hyper* stack, so you can fetch the data from the swarm and leverage local caches. Second, you can optimistically sync the dataset before doing lookups, so it should be possible to get private lookups by downloading more entries than you need.

The registry won't be a blockchain; it will be centrally managed. We're going to make it possible to reconfigure which registry you use so that "forking" is possible, as is layering-in custom/private domain namespaces. Obviously social consensus around names is a tricky topic so "forking the registry" is a complex thing to do, but I think it's important that kind of action is possible.

We're planning to stay compatible with the ICANN domain namespace. To do that, we'll allocate at least one TLD for ourselves (probably `.p2p`) and eventually submit that for membership with ICANN.

For other pre-existing TLDs, we're thinking we can come up with a staged process for integrating them into HDNS.  At first, we'd run HyperDomain tries for the existing TLDs and use some kind of process for registering HyperDomain entries under DNS names you own. That might be something like setting a TXT DNS record under `beakerbrowser.com` proving ownership, and then our HDNS registry would allow you to update the `beakerbrowser.com` HDNS record. Then the plan would be to transfer ownership of those HDNS registries to the orgs that actually control them, if & when they choose to adopt HDNS.

One other thing, though this is still being debated: I'd really like to have an HDNS TLD which is free to register. We'd have to figure out some novel rules for dealing with squatting and the like, but since we expect everybody to be able to create new Hyper* sites for free, I feel like there should be a domain-name space which is also free. (Similar to how everyone expects to publish to NPM under any name without paying.) I'd really prefer that it be a TLD and not a subdomain.

At any rate, I'm looking forward to my `pfrazee.p2p` domain.

**Reply by jtagcat** [reply]
Why not on a blockchain (or append-only log)?

You could make an entry to 'this is my public key', 'this is the domain I'm reserving', 'this is the period I'm reserving it for' (periods capped at say 2mo, renewals can go to 6mo, then 1y, then 3y). Later on 'this is the domain', 'this is the new public key', 'this is my signature of moving the domain'.

Afterwards, 'I'm pointing domain to there', 'this is my signature', with possible overrides.

The only problem I see is spam (reserving every domain). You can't really rate-limit an user, other than making it computationally expensive (the bar needs to be raised, and what about mobile devices). I can suspect some logic with 'max 3 domains pointed at resource', but the resources (hyperdirves, users, identities) can be generated rapidly as well.

The spam problem for users should be addressed anyway.

**Comment by da2x** [comment]
Is there any plans for how existing websites can redirect to a Hyperdrive, though? Say I already rent `example.com` and publish a website on `https://example.com`. E.g. `dat-dns` enables a HTTPS website to auto-discover and redirect to a `dat://hash`.

**Reply by pfrazee** [reply]
We can consider this. Folks: give a thumbs up on my reply here if this interests you. Give a thumbs up to my other reply for the hyper* native approach.

**Reply by Pawper** [reply]
Hmm. It cannot be both?

**Reply by pfrazee** [reply]
Yeah I think that's a possibility

**Reply by voxpelli** [reply]
I like this as a progressive enhancement on top of the existing web. Where you can move from that onto the Hyper-web without having to knowingly make the move. Kind of like how one were upgraded from http to https or http 1 to http 2

**Comment by bndw** [comment]
IIRC I saw a demo using a well-known path on a webserver to jump from `http://` to `hyper://`. @pfrazee is this a thing? If so, does it factor into this?

**Reply by da2x** [reply]
> IIRC I saw a demo using a well-known path on a webserver to jump from `http://` to `hyper://`. @pfrazee is this a thing? If so, does it factor into this?

That’s one of the things dat-dns did.

**Reply by pfrazee** [reply]
I'm looking at a scheme for creating `hyper://` domain entries using HyperDomains rather than the legacy dat-dns hack. I'll describe it in another comment/answer (ctrl+f for HyperDomains).

**Reply by voxpelli** [reply]
@pfrazee Why do you consider it a hack? I felt it was a very good use of the `.well-known` path

**Reply by pfrazee** [reply]
@voxpelli Both of the dat-dns solutions are a little icky. Using DNS TXT records is fine but I dislike using TXT instead of having a real record type. The `.well-known` thing has its upsides but it's a really unusual bootstrap on top of HTTPS which has a ton of overhead.

The main thing that `.well-known` solved for us is dynamic subdomains. HDNS will be able to do that much better, so I'd prefer to drop the well-known. We plan to keep the DNS TXT so that existing domain TLDs can work.

**Reply by voxpelli** [reply]
@pfrazee One good thing with `.well-known` is that it would probably be possible to get that setup on something like a plain GitHub Pages, but yeah, overhead for sure

**Comment by brechtcs** [comment]
I'm more and more of the opinion that legacy DNS is irreparable broken, and that I wouldn't mind if it's merely a second-class citizen in any future Beaker setup. I understand that people want to maintain a link with their `http://` sites (I do too!), but I think for me personally adding the following JavaScript snippet will probably do just fine:

```js
if ('beaker' in window) {
  var redirect = new URL(window.location)
  redirect.protocol = 'hyper:'
  window.location = redirect
}
```

**Reply by bierlingm** [reply]
+1

**Reply by pfrazee** [reply]
I think this is a possibility (hyper:// would basically not use ICANN DNS) but I don't think it's our only option. I'm looking at a scheme for creating hyper:// domain entries using HyperDomains rather than the legacy dat-dns hack. I'll describe it in another comment/answer (ctrl+f for HyperDomains).

**Reply by voxpelli** [reply]
You shouldn't need to execute javascript to be able to jump from one web to another

**Reply by brechtcs** [reply]
That's why I called it a "second class citizen". Also, this discussion has [moved on](https://github.com/beakerbrowser/beaker/discussions/1576#discussioncomment-17437) from this.

**Comment by da2x** [comment]
> I'm more and more of the opinion that legacy DNS is irreparable broken

Any particular reasons? It can be a bit difficult to mange because [it doesn’t follow zeroconf auto-discover contentions](https://github.com/datprotocol/DEPs/issues/45#issue-372678407). But it’s not fundamentally broken.

> but I think for me personally adding the following JavaScript snippet will probably do just fine:

This will hurt performance. DNS lookups are significantly faster than DNS+TLS+HTTP+HTML+JSexec.

**Reply by brechtcs** [reply]
> But it’s not fundamentally broken.

Sorry, I problably wasn't 100% clear on this. I was refering to the current DNS system as a whole, not `dat-dns` in particular. Basically I think the design of DNS is insecure by default, and easily exploited through phishing, man-in-the-middle attacks etc. Our current infrastructure has been forced to accumulate a ton of dirty hacks – such as SSL, HSTS, SFP, DKIM, DMARC – to stop this insecure system from collapsing completely.

So for me the `hyper://` protocol is an opportunity to try and make this entire system obsolete, rather than committing to propping it up indefinitely.

> This will hurt performance. DNS lookups are significantly faster than DNS + TLS + HTTP + HTML + JSexec.

Right, that's what I meant by "second class citizen". Personally, I'm okay with that trade-off.

There's also a slight more performant alternative that requires a bit more work. Beaker currently adds `BeakerBrowser/1.0.0-prerelease.2` to the user agent string. A HTTP server could be configured to detect that and send a 301 response redirecting to `hyper://<key>`. For me, this approach is probably Good Enough™, and I don't think there's a need to broaden the scope of the hyper protocol if a problem like this can be solved in userland.

@pfrazee Maybe it would be good to future-proof this by adding something like `(Hyperdrive support)` to the user agent string? This might be easier for other browsers ([or extensions](https://addons.mozilla.org/en-US/firefox/addon/uaswitcher/)) to adopt as well.

**Reply by pfrazee** [reply]
@brechtcs I'm open to something like that. Is the User Agent the best one to do it? Could also consider a different header, like `Upgrade` or a new header similar to `HTTP2-Settings`

**Reply by brechtcs** [reply]
I like the idea of using an "Upgrade" header, because it's true :)

I don't know the spec that well though, can you just assign it an arbitrary value? If not some servers might return an error when encountering an "Upgrade: hyperdrive" (or whatever value you settle on). But I guess it doesn't really matter whether it's implemented with a header or the user-agent string, both should be quite straightforward to use in an existing HTTP server. Using Nginx for example:

```nginx
server {
  listen 80;
  server_name example.com;
  
  if ($http_upgrade = Hyperdrive) {
    return 301 hyper://<key>
  } 
  # remaining Http config as normal
}
```

Would something like this be acceptable for your use case @da2x?

**Reply by da2x** [reply]
(I’ve already written and submitted this reply but I can’t find it in the thread. Sorry if you guys see it duplicated.)

`Upgrade` is the wrong header for changing the protocol. `Alt-Svc` has everything we need, however. See [RFC-7838](https://tools.ietf.org/html/rfc7838). Chromium already has a system to read and cache this header so that should reduce the implementation time.

    Alt-Svc: hypercore="hyper://hashhash"; ma=15778800
    # service-label=service-uri; max-age=cache-time-in-seconds

The Tor browser uses this to auto-discover .onion addresses on .tld sites and side-grade future requests to the current origin over to the .onion side. So the first page-load happens over HTTPS (everything is already set in motion and the request is being fulfilled.) Future requests are sent to the alternate service instead, though. 

From the user-perspective in the Tor implementation, they see https://example.com even though its served from hashhash.onion. The spec requires the alternate service to have the same HTTPS certificate as the original origin. So the .onion service must have a certificate for example.com.

**We can mitigate that requirement by redirecting from https:// to hyper://.** Changing the origin protocol changes the origin so security is maintained. The auto-discovery happens over a HTTPS connection that is domain-verified by the origin’s certificate. It should be as secure as HTTPS without any MITM vectors.

**Comment by paulharwood** [comment]
There are a number of considerations. Namespaces are a scarce resource and create demand. They are difficult to administer (think visa.p2p). I would ask to possibly rethink this approach.

For me, domain namespacing (dot notation) was always a failure of DNS, ever since inception. DNS failed as content discovery and led to Google and Yahoo! There were many indexing approaches before the oligopolies took over, content discovery was once a flourishing industry of innovation and exploration (webrings anyone!), now it is owned by a select few entities that filter our discovery and ultimately our lives.

The dot notation system is not a natural language. It is a systematic approach to a directory based seek/find problem across a network. Domains used to be physically separated by walls and buildings. 

People don't use dot notation to search (well they kinda did search for shoes.com once upon a time, but direct navigation was killed by Google's omnibox). People certainly don't think that anyone 'owns' a keyword. With hyperdrive, we have a better system now, we have the chance to re-imagine federated search, we can do much better!

I would look at combinations of tags as search terms so rather than: 

`hyper://interesting-topic.p2p`

I would consider using an alphanumeric string, using a `+` as a delimiter and escaping it when necessary. In the UX the `+` could be replaced with a space. This would not conflict with  URI/URL escaping as that is a separate case.

so:

`interesting+topic`

or

`my+interesting+topic`

These terms would defined by the owner of the hyperdrive and could change over time. For example, if you were building a cool app you would call it `cool+app`. The real address scheme could be:

`hyper://[hash]:cool+app:[&query=parameters]`

But let's say your cool app, was not getting discovered, as it was really a todo app with the wrong name, you could change the tags to:

`hyper://[hash]:todo+app:[&query=parameters]`

The key thing to notice here is that the namespace is local and not unique, clashes are expected. Indexing then becomes a trivial local task and you can index based on your own parameters: e.g. how many hops away are the hyperdrives? 1 level, 2 levels, 3 levels deep? A search results page could be crafted to a template and off you go.

Tags as I am proposing are similarly implemented in SSB as Nick Names

https://handbook.scuttlebutt.nz/concepts/identity

Spam becomes a local problem and you can filter based on what you know, what you want. Global hyperdrives are a matter of personal discovery and recommendation. 

It is still a *name system* but not a *domain name system*. You can control the filters.

**Reply by paulharwood** [reply]
I should say that then that typing bare words like

`cool app` 

into the search bar could resolve to 

`hyper://cool+app`

Which could in turn result in a set of results, based on your own search criteria (i.e. proximity)

`hyper://[hash1]:cool+app`
`hyper://[hash2]:cool+app`
`hyper://[hash3]:my+cool+app`

a direct hit to a hyperdrive would still be 

`hyper://[hash]:cool+app`

**Reply by pfrazee** [reply]
@paulharwood A lot of what you're describing is a distributed search system, which I also think is vital but solves a slightly different set of needs. I feel like what we can do is both -- domain names _and_ search -- and then leave users free to not use domain names if it's not their jam.

**Reply by paulharwood** [reply]
Reading the comments below, this discussion has probably already leaked into a bigger domain of expertise and lost focus :)

Yes, you're right, the decision is up to the user ultimately, choice is good. But I'd ask you to consider the paradigm of *search* over *resource descriptors*, which only exists in our collective cognitive behaviour set because existing naming systems are so bad at describing resources. I like the way that SSB/Patchwork allow community based naming, which is not search, but could extend into search.

> In Scuttlebutt, anyone can set a nickname (petname) for another peer. It's akin to a human-memorable symlink to a cryptographic identifier.

We all have the mental model of domains scorched into our foreheads by commercial branding and day to day usage i.e. "amazon*.com*" rather than "amazon". Which is why I accept your point around letting users decide. But even the GTLD space is making that a nonsense now (companyname.blackfriday = $299 a year) and charging a hell of a lot of money for the privilege.

I completely accept that DNS pinning a service is useful, and that this is probably for another discussion, but my point is:

What is more important to describe using a naming system, the hyperdrive _contents_ or the _service_ behind it?

[edit: punctuation]

**Comment by metanomial** [comment]
At peril of perhaps jeopardizing the name security aspect of Zooko's triangle, I really like the idea of scoped namespace registration.

For example, when navigating to `hyper://@mydomain` by clicking on a link or executing JavaScript, Beaker might make the following checks in order:

1. Does the page define the shortname? (Implementation unclear)
2. Does the hyperdrive have a name registry, and does that registry define the shortname?
3. Do any of the client's installed name registries define the shortname?

That way, website operators could ensure that all known links within their hyperdrive always point to the right address. Links created dynamically may still be subject to misresolution.

Beaker could currate and provide a built-in name registry that would be a suitable starting point for most users. More advanced users could supplement or even replace that registry.

**Reply by metanomial** [reply]
Also I agree with the above stated distaste for imitating DNS by using a `p2p` TLD. Besides being familiar and providing an easy way to tell Beaker's search bar that you are looking for a name resolution instead of a search, it's completely unnecessary in this context. I would much rather see a prefix symbol like in my example. Entering an address like `@mydomain/index.html` into the search bar would tell Beaker, "Hey, this is a hyperdrive with shortname `mydomain` and path `/index.html`; Resolve to `hyper://[hash]/index.html`..."

**Reply by pfrazee** [reply]
What's missing from the directory based solution like you describe is an answer for key-rotation. If (as in your example) `@mydomain` only resolves to a hash in specific interfaces like the search-bar, then data or documents (ie HTML) won't be able to leverage that lookup (because they can't support `hyper://@mydomain/` and so their URLs won't update.

Key rotation is a really important property for a system like Hypercore because users have to maintain their own private keys, and we need to expect those to get lost or corrupted.

**Reply by brechtcs** [reply]
> Key rotation is a really important property for a system like Hypercore.

This is something I've been thinking about a bit lately too. I wonder if it might in fact be too important to outsource it to an external solution like DNS or HyperDomains. Could there be a way to address this in `hypercore` itself?

**Reply by pfrazee** [reply]
@brechtcs I don't know how that could be done without requiring users to maintain some kind of master private key, which I don't think is a great solution. I'm going to reply to another comment of yours with an in-depth explanation for why I do see HDNS as strong.

**Reply by serapath** [reply]
I do feel like making users maintain a master private key is workable.
There are a lot of youtube videos about `key ceremonies` and how to backup a **mnemonic** based on https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki ... the whole thing is, that if a user forgets their keys, they are as well off as users who don't have a master key in the first place, but if they actually have a master key, they or somebody whom they trust could still one day decide to override and opt-out of maybe some kind of authority that otherwise signs on behalf of them.

**Reply by DougAnderson444** [reply]
I agree with the master private key, as long as backups/recovery methods are built in. You can do Shamir Secret Share with the bip39 mnemonic and give the shares to three hyper-friends as a backup backup.

**Comment by da2x** [comment]
I change the topic to narrow the scope a bit. There’s a lot of interesting ideas but little is about the original topic. The top post is about mapping ICANN DNS to Hyperdrives. This is a feature that exists in Dat to facilitate seamless migration from a HTTPS website to a P2P equivalent of that same website.

I believe it’s important to drive adoption among existing web-enthusiasts who’ve invested in their existing domains. Plus, you know, hashes aren’t meant for humans. DNS domains are.

**Reply by pfrazee** [reply]
To answer your question: Yes, after more recent conversations, we're planning to use DNS TXT records for existing ICANN TLDs. HDNS will operate as a separate solution for its own TLDs.

**Reply by paulharwood** [reply]
I agree, scope is important. But it is an opportunity to highlight the issue of using authoritative naming systems to interface with peer based, distributed filesystems.

**Reply by brechtcs** [reply]
> Plus, you know, hashes aren’t meant for humans.

No argument there.

> DNS domains are.

But I'm not sure I agree with this. The large majority of non-technical user rarely use domain names for navigation. From my observations, the most common usage is a bit as follows:

1. Type the title of the desired site or page in the URL bar
2. Check autocomplete results to see if the browser already found it its history or bookmarks
3. If it's not in autocomplete, just press Enter to find it on Google

Steps 1 and 2 already work perfectly on Beaker without any need for domain names. Step 3 will hopefully be solved in userland sometime soon, and doesn't require DNS support either.

**Reply by pfrazee** [reply]
@brechtcs I agree that distributed search is important because it does provide a fully decentralized answer to lookup (and is often a good UX). I don't agree that domains are not user-friendly; it's true many people use search, but I think domain names have had a lot of good utility.

I also think it's important to reiterate, you won't have to use domain names if you don't want to.

I do think HDNS is a notable improvement to existing DNS. Here's why:

 - The Hypertrie datastructure which HDNS will use is externally auditable (as all Hypercore data structures are). Every change to a domain-mapping is publicly broadcast and can't be altered without appearing in the changelog. (The entire system is extremely similar to Certificate Transparency.)
 - You can't MITM HDNS to alter the response (as with all Hypercores).
 - Because of the auditability, you don't need external CAs to verify the mapping.
 - You can optimistically sync more of the HDNS registries than needed and therefore avoid leaking your domain lookups.
 - HDNS will have a much stronger API than DNS for a two reasons. 1. The TLD registries will be designed with a standard Web API for managing domains. 2. The subdomains will be delegated to Hypertries that users create. The delegated-trie model means that you can update your subdomains by just writing to your local HDNS structure and syncing it out.

Regarding centralization of the registry management, I think there are a couple of ways we can help that situation. The first thing we can do is making the TLD registries configurable. I see that as a similar solution to forking in FOSS: it gives users a way to overthrow the current leaders if needed. The second thing we can do is make it possible to override individual entries, or perhaps even layer multiple registries under a TLD. That way users can make non-global or personal name mappings.

**Reply by paulharwood** [reply]
Yes, domains are useful to describe the location of a resource, but are bad at describing a service. 

e.g. take `maps` as a search term. If I manage to get hold of `maps.p2p` and just put car adverts on it, that is a failed service descriptor, but useful for me if I want to be discovered as `paul@maps.p2p`

Personally, i'd love to see the typing of `maps` into the search bar would give me a search results list that I could format as I see fit. Like a google search page that you can customise. 

`hyper://openstreetmap.com:maps/`
`hyper://@openstreetmap:maps/`
`hyper://@pfrazee:maps/`
`hyper://@paulharwood:world+maps/`

[edit: missing slashes, removed conjecture]

**Reply by serapath** [reply]
I second that.
This is how these systems are actually used in practice already.
1. type into the search bar
2. choose from autocomplete options and if that doesnt work
3. hit enter and search to find the list on google and click

The choices you make most often will show up more towards the top over time.

---

1. You might not be able to find certain entries at all
2. but you might be able - similar to the weight http://node-modules.com/search?q=beaker would customize a search for npm modules too

Of course - if it really doesn't work, you might try different kinds of searches or use
1. either names of peers on the edges of your social graph
2. or hyper addresses of hopefully related content
to get better search results, or - in the end - you can always find somebody who gives you a direct hyperaddress link.

---

Lastly, if you want to share a link and express your opinion or the opinion about the name of a hyper address you got from somebody else giving you that link - like even your autocomplete search bar could recommend you and offer you copy/paste, so that you would essentially pass on something like:

`hyper://1c6d8c9e2bca71b63f5219d668b0886e4ee2814a818ad1ea179632f419ed29c4:foo+bar+baz`

If another user was opening that link, their local browser would already know `foo+bar+baz` to be one of the words it should associate with that address

**Comment by da2x** [comment]
`Upgrade` is the wrong header for this purpose. However, `Alt-Svc` fits the bill. See [RFC 7838](https://tools.ietf.org/html/rfc7838).

Example:

    Alt-Svc: hypercore="hyper://hashash"; ma=2592000; persist=1
    # (service-label=origin-uri, max-age=cache-in-seconds, persist-in-cache=true)

The Tor browser uses this header to auto-discover and serve traffic from a .onion address on a .TLD. The browser still shows example.com even though future requests are loaded over example.onion. However, this header could be used to redirect from https to hyper.

The HTTPS server sends this header over HTTPS so there's is some built in security. The spec requires the Alt-Svc target to be served from a server with s certificate for the primary origin. Redirecting to a different protocol mitigates the underlaying requirements for this, however.

**Reply by paulharwood** [reply]
Awesome, this looks like a fantastic option. Maybe this is the opportunity to define the schema and pin the services together in a compatible way?

`Alt-Svc: hypercore="hyper://@paulharwood:world+maps"; ma=2592000; persist=1`

Resolves to/instead of:

`Alt-Svc: hypercore="hyper://f7827a05090cdc7982910c167accd991be98eda35ad1253f946728768386cd2e"; ma=2592000; persist=1`

**Comment by da2x** [comment]
Test. (My replies to this thread keeps disappearing.)

**Reply by paulharwood** [reply]
Mine too

**Comment by da2x** [comment]
This is so weird. I get emails about new updates to this thread but I can’t see anything on GitHub. Maybe we need a Hyperdrive-powered gossip discussion board sooner rather than later.

**Comment by gigablast** [comment]
Ok, very interesting stuff here. Being a search guy here's my specific take on it. I'm not sure i know wtf i am talking about. But i am assuming there is some central 'swarm file' that maps a hyperdrive key to an IP address:
1) include the hyperdrive *title* in the swarm file (possibly truncated)
2) allow all hyperdrives (unique IP C-blocks) to vote on other hyperdrives that they think are good. include these votes in the swarm file, too.
3) then a user can type in query like 'maps' and find the highest rated hyperdrives with 'maps' (or synonym thereof) in the title. [title keyword density plays a role of course, too!]

so it's some sort of "decentralized democractic dns" quite similar to a lot of the other decrentralized approaches.

**Reply by gigablast** [reply]
*and people can downvote certain IPs and hyperdrives too, to help combat spam and spammy voters.

**Reply by gigablast** [reply]
*additionally, voters could also supply keywords as to what they think best represent the hyperdrive they are voting on. and these keywords would be searched as if they were part of the title of the hyperdrive being voted on.

**Reply by gigablast** [reply]
*lastly, you could restrict voters to a group of voters you trust. this would be similar to ICANN if ICANN consisted of a bunch of voters. that would make things even more immune to spam. like, several groups of people/voters would emerge as trustworthy, but perhaps differ in political views. and you could pick one to use when doing your searches.

**Comment by gigablast** [comment]
it seems you already have a list of hyperdrives on a single page from your list.json file. this can be the dns lookup for now. :)

**Comment by RangerMauve** [comment]
For what it's worth, I'm still supporting dat-dns inside [Agregore](https://github.com/RangerMauve/agregore-browser/) and it's working great for my blog: hyper://blog.mauve.moe 

I've also got it seamlessly working with [dat-store](https://github.com/datproject/dat-store) where I have it keeping a backup of my archive and serving as an HTTP gateway for any archives I store there. Then I use NGINX for setting up letsencrypt certificates and have it redirecting to the gateway URL.

**Comment by afefafafe** [comment]
Has there been any updates for something like this?

**Reply by pfrazee** [reply]
I think we're going to stick with the dat-dns solution for now (though we should talk about whether the DNS TXT record should change [per some issues raised](https://github.com/datprotocol/DEPs/issues/45)). As much as we like the hyper-dns idea, there's a complex social issue around appointing central-writer authority which has to be resolved: creating a market of registrars. ICANN solves that issue by artificially creating one -- they act as the coordinating point between accredited registrars but don't accept registrations themselves. Unless we did the same, or used a completely ownerless market (ya know, blockchains) then we'd be setting a bad precedent for the market by running the sole registrar. So, unless & until we can solve that, we're going to piggyback off existing DNS.

**Comment by HughIsaacs2** [comment]
Something about DNS that I felt was an issue with the Dat versions of Beaker Browser, was that a Dat loaded via its key and that same Dat loaded via a domain name would be treated as two different origins.

I've always thought DNS should just act as a mask for the regular Dat URL.

Could that be applied to hyper:// or is there something valid that I'm missing? (security concerns maybe?)

**Reply by RangerMauve** [reply]
I think the main issue there is that Electron doesn't provide a way for us to specify the origin to be something other than what's in the URL.

In the past I considered redirecting to the public key version when you access the DNS version, but that can be a problem too because sometimes you'd want the flexibility of being able to change what the DNS points to.

**Reply by HughIsaacs2** [reply]
I was thinking not even using Electron's own features for this and instead just having the address bar display the domain name while the webview loads the public key as normal.
Only supporting Electron's features for cases where a site might want to load something from the domain name and not specify the specific key.

**Reply by RangerMauve** [reply]
Yeah, that's a good idea for top level navigation. I wonder if it'd matter for when you have iframes or other resources linked to from within the page. 🤔
